cadc-pkg-server/src/main/java/org/opencadc/pkg/server/PackageRunner.java

```diff
@@ -241,31 +233,49 @@ private void doIt() {
     }
 
     /**
-     * Initialize syncOutput output stream with the correct content type and disposition,
-     * as provided by the writer class
-     * @param packageName - package name to use
+     * Set up PackageWriter and syncoutput for the requested RESPONSEFORMAT value.
+     * Initialize syncOutput output stream with the correct content type and disposition
+     * as provided by the writer class. Call writer ctor.
+     * @param packageName
+     * @return
      * @throws IOException
      */
-    private void initOutputStream(String packageName) throws IOException {
+    private PackageWriter initWriter(String packageName) throws IllegalArgumentException, IOException {
 
         if (!StringUtil.hasText(packageName)) {
-            throw new RuntimeException("BUG: packageName can't be null");
+            throw new IllegalArgumentException("packageName not provided");
         }
 
-        syncOutput.setResponseCode(200);
+        // Package type is in RESPONSEFORMAT Job parameter
+        String responseFormat = ParameterUtil.findParameterValue("RESPONSEFORMAT", job.getParameterList());
+
+        if (!StringUtil.hasLength(responseFormat)) {
+            throw new IllegalArgumentException("RESPONSEFORMAT not found in job");
+        } else {
+
+            syncOutput.setResponseCode(200);
+
+            StringBuilder cdisp = new StringBuilder();
+            cdisp.append("inline;filename=");
+            cdisp.append(packageName);
 
-        // TODO: when a mechanism to select the package type
-        // is included in PackageRunner, the content type and
-        // disposition values will depend on values set on entry to
-        // PackageRunner. For now they're hard coded for tar files.
-        StringBuilder cdisp = new StringBuilder();
-        cdisp.append("inline;filename=");
-        cdisp.append(packageName);
-        cdisp.append(".tar");
+            if (responseFormat.equals(ZipWriter.MIME_TYPE)) {
+                cdisp.append(".zip");
+                syncOutput.setHeader("Content-Type", ZipWriter.MIME_TYPE);
+                syncOutput.setHeader("Content-Disposition", cdisp.toString());
+                this.outputStreamCounter = new ByteCountOutputStream(syncOutput.getOutputStream());
+                return new TarWriter(this.outputStreamCounter);
+
+            } else if (responseFormat.equals(TarWriter.MIME_TYPE)) {
+                cdisp.append(".tar");
+                syncOutput.setHeader("Content-Type", TarWriter.MIME_TYPE);
+                syncOutput.setHeader("Content-Disposition", cdisp.toString());
+                this.outputStreamCounter = new ByteCountOutputStream(syncOutput.getOutputStream());
```

Comment:
The above 2 lines (and in the block above) can be brought out because they are the same in both.

Reply:
Reworked this section a bit. renamed 'outputStreamCounter' to 'bcOutputStream' - the 'Counter' part was a misnomer. It's used in run() to log the amount of bytes streamed.